File: node/src/rules.js
const fs = require("fs");
const path = require("path");

const RULE_FILES = ["characters", "typography", "phrases", "homoglyphs"];

function candidateDirs() {
  const dirs = [];
  if (process.env.WMC_RULES_DIR) dirs.push(process.env.WMC_RULES_DIR);
  dirs.push(path.join(__dirname, "..", "rules_data"));
  let current = __dirname;
  for (let i = 0; i < 6; i += 1) {
    dirs.push(path.join(current, "rules"));
    current = path.dirname(current);
  }
  return dirs;
}

function resolveDir() {
  for (const dir of candidateDirs()) {
    if (fs.existsSync(path.join(dir, "characters.json"))) return dir;
  }
  throw new Error("wmc rules not found. set WMC_RULES_DIR or run from the repository.");
}

let cache = null;

function loadRules() {
  if (cache) return cache;
  const base = resolveDir();
  const rules = {};
  for (const name of RULE_FILES) {
    rules[name] = JSON.parse(fs.readFileSync(path.join(base, `${name}.json`), "utf-8"));
  }
  cache = rules;
  return rules;
}

module.exports = { loadRules };

File: python/wmc/rules.py
import json
import os
from functools import lru_cache
from pathlib import Path

_RULE_FILES = ["characters", "typography", "phrases", "homoglyphs"]


def _candidate_dirs():
    env = os.environ.get("WMC_RULES_DIR")
    if env:
        yield Path(env)
    here = Path(__file__).resolve()
    yield here.parent / "rules_data"
    for parent in here.parents:
        candidate = parent / "rules"
        if candidate.is_dir():
            yield candidate


def _resolve_dir():
    for candidate in _candidate_dirs():
        if candidate.is_dir() and (candidate / "characters.json").is_file():
            return candidate
    raise FileNotFoundError(
        "wmc rules not found. set WMC_RULES_DIR or run from the repository."
    )


@lru_cache(maxsize=1)
def load_rules():
    base = _resolve_dir()
    rules = {}
    for name in _RULE_FILES:
        with open(base / f"{name}.json", "r", encoding="utf-8") as handle:
            rules[name] = json.load(handle)
    return rules
